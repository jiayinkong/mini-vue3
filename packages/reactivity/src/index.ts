export { ref, unRef, isRef, proxyRefs } from './ref';
export { reactive, readonly, isReactive, isReadonly, shallowReadonly, isProxy } from './reactive';
export { effect, stop, ReactiveEffect } from './effect';