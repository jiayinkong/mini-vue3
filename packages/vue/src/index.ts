export * from '@mini-vue3/runtime-dom';

import { baseCompile } from '@mini-vue3/compiler-core';
import * as runtimeDom from '@mini-vue3/runtime-dom';
import { registerRuntimeCompiler } from '@mini-vue3/runtime-dom';

function compileToFunction(template) {
  const { code } = baseCompile(template);
  const render = new Function('Vue', code)(runtimeDom);
  return render;
}

registerRuntimeCompiler(compileToFunction);


          // compiler-core
// Vue ->     
          // runtime-dom -> runtime-core -> reactivity