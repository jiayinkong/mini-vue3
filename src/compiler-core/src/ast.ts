import { CREATE_ELEMENT_VNODE } from './runtimeHelpers';

export const enum NodeTypes {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT,
  TEXT,
  ROOT,
  COMPOUND,
}

export const enum TagType {
  Start,
  End,
}

export function createVNodeCall(context, tag, props, children) {
  context.helper(CREATE_ELEMENT_VNODE);
  return {
    type: NodeTypes.ELEMENT,
    tag,
    children,
    props,
}
}