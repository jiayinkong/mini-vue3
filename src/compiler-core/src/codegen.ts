export function generate(ast) {
  const context = createCodegenContext();
  let { push } = context;

  let functionName = 'render';
  let args = ['_ctx', '_cache'];
  let signature = args.join(', ');

  push(`function ${functionName}(${signature}){`);

  push('return ');
  getNode(ast.codegenNode, context);
  push('}');

  return {
    code: context.code,
  }
}

function createCodegenContext() {
  const context = {
    code: '',
    push(source) {
      context.code += source;
    }
  }

  return context;
}

function getNode(node, context) {
  const { push } = context;
  push(`'${node.content}'`);
}