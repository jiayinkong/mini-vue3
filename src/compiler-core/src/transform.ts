export function transform(root, options = {}) {
  const context = createTransformText(root, options);

  // 1.遍历 - 深度优先遍历
  traverseNode(root, context);

  createRootCodegen(root);
}

function createRootCodegen(root) {
  return root.codegenNode = root.children[0];
}

function createTransformText(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
  };

  return context;
}

function traverseNode(node, context) {
  console.log(node);

  const nodeTransforms = context.nodeTransforms;
  for(let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i];
    transform(node);
  }

  const children = node.children;

  if(children) {
    for(let i = 0; i < children.length; i++) {
      const node = children[i];
      traverseNode(node, context);
    }
  }
}