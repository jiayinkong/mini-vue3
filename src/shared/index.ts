export * from './toDisplayString';

export const extend = Object.assign;

export const isObject = (val) => {
  return val !== null && typeof val === 'object';
}

export const isString = (val) => {
  return typeof val === 'string';
};

export const hasChanged = (val, newVal) => {
  return !Object.is(val, newVal);
}


// 添加 on 前缀，并事件首字母大写
const capitalize = (str: string) => {
  return str ? str.charAt(0).toUpperCase() + str.slice(1) : '';
}

// 改为驼峰
export const camelize = (str: string) => {
  return str.replace(/-(\w)/g, (_, c: string) => {
    return c ? c.toUpperCase() : '';
  })
}

export const toHandlerKey = (str) => {
  return str ? 'on' + capitalize(str) : '';
}
  