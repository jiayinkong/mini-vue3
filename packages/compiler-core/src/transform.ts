import { NodeTypes } from './ast';
import { TO_DISPLAY_STRING } from './runtimeHelpers';

export function transform(root, options = {}) {
  const context = createTransformText(root, options);

  // 1.遍历 - 深度优先遍历
  traverseNode(root, context);

  createRootCodegen(root);

  root.helpers = [...context.helpers.keys()];
}

function createRootCodegen(root) {
  root.codegenNode = root.children[0];
  const child = root.children[0];
  if(child.type === NodeTypes.ELEMENT) {
    root.codegenNode = child.codegenNode;
  } else {
    root.codegenNode = root.children[0];
  }
}

function createTransformText(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1);
    }
  };

  return context;
}

function traverseNode(node, context) {
  const nodeTransforms = context.nodeTransforms;
  let existFns: any = [];

  for(let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i];
    const onExist = transform(node, context);
    if(onExist) existFns.push(onExist);
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING);
      break;

    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChildren(node, context);
      break;
  
    default:
      break;
  }

  let i = existFns.length;
  while(i--) {
    existFns[i]();
  }
}

function traverseChildren(node, context) {
  const children = node.children;

  for(let i = 0; i < children.length; i++) {
    const node = children[i];
    traverseNode(node, context);
  }
}