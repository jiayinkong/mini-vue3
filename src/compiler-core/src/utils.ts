import { NodeTypes } from './ast';

export function isText(child) {
  return child.type === NodeTypes.TEXT || child.type === NodeTypes.INTERPOLATION;
}