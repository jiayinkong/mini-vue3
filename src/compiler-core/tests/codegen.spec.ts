import { generate } from '../src/codegen';
import { baseParse } from '../src/parse'
import { transform } from '../src/transform';
import { transformExpression } from '../src/transforms/transformExpression';

describe('generate', () => {
  it('string', () => {
    const ast = baseParse('hiiii');

    transform(ast);

    const { code } = generate(ast);
  
    expect(code).toMatchSnapshot();
  });

  it('inperpolation', () => {
    const ast = baseParse('{{message}}');

    transform(ast, {
      nodeTransforms: [transformExpression]
    });

    const { code } = generate(ast);
  
    expect(code).toMatchSnapshot();
  });
})