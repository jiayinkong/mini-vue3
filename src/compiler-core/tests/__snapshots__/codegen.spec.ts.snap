// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate element 1`] = `
"const { toDisplayString:  _toDisplayString, createElementVNode:  _createElementVNode } = Vue
return function render(_ctx, _cache){return _createElementVNode('div', null, 'Hi, ' + _toDisplayString(_ctx.message))}"
`;

exports[`generate inperpolation 1`] = `
"const { toDisplayString:  _toDisplayString } = Vue
return function render(_ctx, _cache){return _toDisplayString(_ctx.message)}"
`;

exports[`generate string 1`] = `
"
return function render(_ctx, _cache){return 'hiiii'}"
`;
